File: src/wishCompiler.ts
import {
  ProductSignal,
  Wish,
  WishStory,
  WishTask,
} from './types';

// ============================================================================
// WISH COMPILER v0.1
// ============================================================================
// Read this file from top to bottom.
// It is intentionally plain TypeScript.
//
// The basic idea:
//
// HUMAN WISH
//    -> TASK
//    -> STORY
//    -> PRODUCT SIGNALS
//    -> WISH WELL ROUTING
//    -> HUMAN-GOVERNED ACTION
//
// This file does NOT make legal, medical, financial, insurance, or safety
// decisions. It only structures the wish and flags when human review is needed.

const HIGH_STAKES_WORDS = [
  'legal',
  'court',
  'custody',
  'medical',
  'therapy',
  'insurance',
  'safety',
  'violence',
  'abuse',
  'benefits',
  'credit',
  'tax',
];

// A simple lookup table connects ordinary language to WISH WELL houses.
// Later, AI can help classify a wish, but the routing rules should remain
// visible and reviewable instead of becoming mysterious model behavior.
const HOUSE_KEYWORDS: Record<number, string[]> = {
  1: ['identity', 'family', 'business', 'entity'],
  2: ['money', 'budget', 'income', 'asset', 'price', 'financial'],
  3: ['technology', 'communication', 'email', 'phone', 'ai'],
  4: ['home', 'house', 'housing', 'health', 'food', 'care'],
  5: ['project', 'build', 'create', 'product', 'play'],
  6: ['system', 'habit', 'routine', 'process', 'workflow'],
  7: ['contract', 'agreement', 'accountability', 'legal'],
  8: ['risk', 'insurance', 'safety', 'protection', 'compliance'],
  9: ['travel', 'therapy', 'education', 'school', 'research', 'trust'],
  10: ['story', 'legacy', 'press', 'public', 'authority'],
  11: ['community', 'partner', 'network', 'sponsor'],
  12: ['mind', 'body', 'soul', 'regulation', 'meditation'],
  13: ['creation', 'template', 'recipe', 'deal', 'alchemy'],
};

function normalize(text: string): string {
  return text.trim().toLowerCase();
}

export function routeWishToHouses(statement: string): number[] {
  const text = normalize(statement);
  const houses: number[] = [];

  for (const [houseNumber, keywords] of Object.entries(HOUSE_KEYWORDS)) {
    const matchesHouse = keywords.some((keyword) => text.includes(keyword));

    if (matchesHouse) {
      houses.push(Number(houseNumber));
    }
  }

  // Every wish is ultimately part of the wish-to-deal / creation process.
  if (!houses.includes(13)) {
    houses.push(13);
  }

  return houses;
}

export function requiresHumanReview(statement: string): boolean {
  const text = normalize(statement);

  return HIGH_STAKES_WORDS.some((word) => text.includes(word));
}

export function createTask(statement: string, desiredState: string): WishTask {
  return {
    id: `task-${Date.now()}`,
    action: `Create an executable plan for: ${desiredState}`,
    dependencies: [
      'confirm the current state',
      'identify constraints',
      'identify people and resources involved',
    ],
    status: 'proposed',
  };
}

export function createStory(
  statement: string,
  currentState: string,
  desiredState: string,
): WishStory {
  return {
    actor: 'the person or household making the wish',
    situation: currentState,
    need: statement,
    desiredOutcome: desiredState,
    barriers: [],
    acceptanceCriteria: [
      'the desired state is specific enough to evaluate',
      'the next actions are understandable to the user',
      'high-stakes actions remain under human or licensed-professional review',
    ],
  };
}

export function identifyProductSignals(story: WishStory): ProductSignal[] {
  // In v0.1, we create one transparent signal from the story.
  // Later this can expand into several signals or use an AI model with a
  // structured schema and human approval.
  return [
    {
      problem: `A gap exists between "${story.situation}" and "${story.desiredOutcome}".`,
      affectedUsers: [story.actor],
      reusable: true,
      evidenceNeeded: [
        'user confirmation',
        'barriers encountered during execution',
        'outcome after the task is completed',
      ],
    },
  ];
}

export function compileWish(input: {
  id: string;
  ownerId: string;
  statement: string;
  currentState: string;
  desiredState: string;
  source?: string;
}): Wish {
  const task = createTask(input.statement, input.desiredState);
  const story = createStory(
    input.statement,
    input.currentState,
    input.desiredState,
  );

  const humanReviewRequired = requiresHumanReview(input.statement);

  return {
    id: input.id,
    ownerId: input.ownerId,
    statement: input.statement,
    currentState: input.currentState,
    desiredState: input.desiredState,
    source: input.source ?? 'manual',
    createdAt: new Date().toISOString(),
    houses: routeWishToHouses(input.statement),
    tasks: [task],
    story,
    productSignals: identifyProductSignals(story),
    approvalRequired: humanReviewRequired,
    humanReviewRequired,
    status: 'captured',
  };
}

// ---------------------------------------------------------------------------
// EXAMPLE
// ---------------------------------------------------------------------------
// This object is here so you can understand the code without needing the UI.
// Nothing happens automatically just because this example exists.

export const EXAMPLE_WISH = compileWish({
  id: 'wish-example-001',
  ownerId: 'example-user',
  statement: 'I wish I could relocate my family and keep school, housing, money, and healthcare stable.',
  currentState: 'The household is preparing for a major transition.',
  desiredState: 'The family is established in a stable new home with essential services connected.',
  source: 'cookbook',
});

File: learning/wish_compiler.py
"""WISH WELL Wish Compiler - Python learning version.

This file mirrors the ideas in src/wishCompiler.ts using beginner Python.
Run it directly and read it from top to bottom.

WISH -> TASK -> STORY -> PRODUCT SIGNAL -> ACTION -> OUTCOME -> RECIPE

This is a learning prototype. It does not make legal, medical, financial,
insurance, custody, or safety decisions. It only structures information and
flags subjects that require human review.
"""


HIGH_STAKES_WORDS = [
    "legal",
    "court",
    "custody",
    "medical",
    "therapy",
    "insurance",
    "safety",
    "violence",
    "abuse",
    "benefits",
    "credit",
    "tax",
]


HOUSE_KEYWORDS = {
    1: ["identity", "family", "business", "entity"],
    2: ["money", "budget", "income", "asset", "price", "financial"],
    3: ["technology", "communication", "email", "phone", "ai"],
    4: ["home", "house", "housing", "health", "food", "care"],
    5: ["project", "build", "create", "product", "play"],
    6: ["system", "habit", "routine", "process", "workflow"],
    7: ["contract", "agreement", "accountability", "legal"],
    8: ["risk", "insurance", "safety", "protection", "compliance"],
    9: ["travel", "therapy", "education", "school", "research", "trust"],
    10: ["story", "legacy", "press", "public", "authority"],
    11: ["community", "partner", "network", "sponsor"],
    12: ["mind", "body", "soul", "regulation", "meditation"],
    13: ["creation", "template", "recipe", "deal", "alchemy"],
}


def normalize(text):
    """Make text easier to compare."""
    return text.strip().lower()


def route_wish_to_houses(statement):
    """Return WISH WELL houses whose visible keywords match the Wish."""
    text = normalize(statement)
    houses = []

    # A loop repeats the same check for every house.
    for house_number, keywords in HOUSE_KEYWORDS.items():
        # any(...) becomes True if at least one keyword appears in the text.
        matches_house = any(keyword in text for keyword in keywords)

        if matches_house:
            houses.append(house_number)

    # House 13 represents the wish-to-deal / creation process.
    if 13 not in houses:
        houses.append(13)

    return houses


def requires_human_review(statement):
    """Flag high-stakes subject matter without deciding the underlying issue."""
    text = normalize(statement)
    return any(word in text for word in HIGH_STAKES_WORDS)


def create_task(statement, desired_state):
    """Turn the Wish into proposed executable work."""
    return {
        "action": f"Create an executable plan for: {desired_state}",
        "source_wish": statement,
        "dependencies": [
            "confirm the current state",
            "identify constraints",
            "identify people and resources involved",
        ],
        "status": "proposed",
    }


def create_story(statement, current_state, desired_state):
    """Turn the task into human context for product development."""
    return {
        "actor": "the person or household making the wish",
        "situation": current_state,
        "need": statement,
        "desired_outcome": desired_state,
        "barriers": [],
        "acceptance_criteria": [
            "the desired state is specific enough to evaluate",
            "the next actions are understandable to the user",
            "high-stakes actions remain under human review",
        ],
    }


def identify_product_signals(story):
    """Turn the gap in the story into reusable product-development evidence."""
    return [
        {
            "problem": (
                f'A gap exists between "{story["situation"]}" and '
                f'"{story["desired_outcome"]}".'
            ),
            "affected_users": [story["actor"]],
            "reusable": True,
            "evidence_needed": [
                "user confirmation",
                "barriers encountered during execution",
                "outcome after the task is completed",
            ],
        }
    ]


def compile_wish(wish_id, owner_id, statement, current_state, desired_state):
    """Run the smaller recipes and return one structured Wish dictionary."""
    task = create_task(statement, desired_state)
    story = create_story(statement, current_state, desired_state)
    human_review_required = requires_human_review(statement)

    wish = {
        "id": wish_id,
        "owner_id": owner_id,
        "statement": statement,
        "current_state": current_state,
        "desired_state": desired_state,
        "houses": route_wish_to_houses(statement),
        "tasks": [task],
        "story": story,
        "product_signals": identify_product_signals(story),
        "approval_required": human_review_required,
        "human_review_required": human_review_required,
        "status": "captured",
    }

    return wish


def explain_wish(wish):
    """Print the important parts in human language."""
    print("\nWISH")
    print(wish["statement"])

    print("\nDESIRED STATE")
    print(wish["desired_state"])

    print("\nWISH WELL HOUSES")
    print(wish["houses"])

    print("\nTASK")
    print(wish["tasks"][0]["action"])

    print("\nSTORY")
    print(wish["story"])

    print("\nPRODUCT SIGNAL")
    print(wish["product_signals"][0]["problem"])

    print("\nHUMAN REVIEW REQUIRED")
    print(wish["human_review_required"])


if __name__ == "__main__":
    # Change these three lines first when you practice.
    example_wish = compile_wish(
        wish_id="wish-example-001",
        owner_id="example-user",
        statement=(
            "I wish I could relocate my family and keep school, housing, "
            "money, and healthcare stable."
        ),
        current_state="The household is preparing for a major transition.",
        desired_state=(
            "The family is established in a stable new home with essential "
            "services connected."
        ),
    )

    explain_wish(example_wish)
